docs/extensibility/parser.md

```diff
@@ -0,0 +1,36 @@
+# Extending the Parser
+
+When the editor is interacting with blocks, these are stored in memory as data structures comprising a few basic properties and attributes. Upon saving a working post we serialize these data structures into a specific HTML structure and save the resultant string into the `post_content` property of the post in the WordPress database. When we load that post back into the editor we have to make the reverse transformation to build those data structures from the serialized format in HTML.
+
+The process of loading the serialized HTML into the editor is performed by the _block parser_. The formal specification for this transformation is encoded in the parsing expression grammar (PEG) inside the `@wordpress/block-serialization-spec-parser` package. The editor provides a default parser implementation of this grammar but there may be various reasons for replacing that implementation with a custom implementation. We can inject our own custom parser implementation through the appropriate filter.
+
+## Server-side parser
+
+Plugins have access to the parser if they want to process posts in their structured form instead of a plain HTML-as-string representation. WordPress itself uses the server-side parser when rendering pages: it removes the block comment delimiters and it replaces dynamic blocks with their rendered content.
```

Comment:
> WordPress itself uses the server-side parser when rendering pages: it removes the block comment delimiters and it replaces dynamic blocks with their rendered content.

We're not there yet, we should remove this line.

Reply:
not there? last time I checked we were. the server-side parser is currently a partial implementation right? I thought for sure that we both had access to the full parse on the server (for plugins) _and_ WordPress itself was parsing parts of structure to remove comments and render dynamic blocks - not a full tree, but top-level blocks

Reply:
We do process dynamic blocks, but we don't use the server-side parser (meaning `gutenberg_parse_blocks`) in `do_blocks` since #4591. I suggested removing this line because it sounds like the same (full) parser is being used for dynamic blocks.

After the current PR, I'm sure we'll be able to switch to just using the new parser, and update the documentation here.

Reply:
removed in 2e4fc348a